### PR TITLE
backend/feat: 873 rideForceSync api

### PR DESCRIPTION
**Pull request by roman-bodavskiy**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
We’ll create another version of rideSync called rideForceSync which will just flush the status from BPP to BAP.
We will continue to use rideSync first, and run the force version on failures from the first attempt.
We give an ID to BPP, BPP sends that ID’s BPP status to BAP, and BAP simply updates it.
BPP doesn’t need to know BAP’s status. BAP gets BPP’s status and consumes it.

Additional changes

- **bap on_update**: If `target status` is already updated in `target location` for one of `Booking/Ride`, it should not prevent the other being updated to `target status`. (implemented for `RideCompleted` and `RideCancelled` updates)

- **bpp rideSync**: the cancellation also fails if cancellation reason is not provided in the BPP side. That shouldn’t be an error. Only warning. The operation should go through.

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## How did you test it?
Tested locally

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/dhall-configs/dev/dynamic-offer-driver-app.dhall [hunk 1]
```diff
@@ -13,7 +13,7 @@ let esqDBCfg =
 
 let esqDBReplicaCfg =
       { connectHost = esqDBCfg.connectHost
-      , connectPort = 5435
+      , connectPort = 5433
```
Comment (NikithShetty) [inline Backend/dhall-configs/dev/dynamic-offer-driver-app.dhall:16]: @roman-bodavskiy why is this changed?

**Comment by NikithShetty** [comment]
@roman-bodavskiy This PR implements force sync as a BPP side push with `force_update` tag in the Beckn API.
I don't think this is the right approach.

Its not the responsibility of the BPP to keep the BAP in sync. IMO This should be a BAP side functionality, where BAP should call `/status` API of BPP and sync its internal data models based on `/on_status` API call from BPP.

Normal sync API should also have been implemented this way.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnStatus.hs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import qualified Domain.Action.Beckn.Status as DStatus
+import qualified Domain.Types.Ride as DRide
+import Kernel.Types.Id (Id)
+
+mkOnStatusMessage :: Id DRide.Ride -> DStatus.DStatusRes -> OnStatus.OnStatusMessage
+mkOnStatusMessage rideId res = do
+  OnStatus.OnStatusMessage
+    { order =
+        OnStatus.Order
+          { id = rideId.getId,
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnStatus.hs:27]: order.id is bookingId and fulfilmentId is rideId

Comment (RitikaHanish) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnStatus.hs:27]: @olegromashin Can you do these changes ?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 3]
```diff
@@ -243,22 +251,25 @@ onUpdate RideCompletedReq {..} = do
           }
 onUpdate BookingCancelledReq {..} = do
   booking <- QRB.findByBPPBookingId bppBookingId >>= fromMaybeM (BookingDoesNotExist $ "BppBookingId: " <> bppBookingId.getId)
-  unless (isBookingCancellable booking) $
-    throwError (BookingInvalidStatus (show booking.status))
   mbRide <- QRide.findActiveByRBId booking.id
+
+  let isBookingCancellable =
+        booking.status `elem` [SRB.NEW, SRB.CONFIRMED, SRB.AWAITING_REASSIGNMENT, SRB.TRIP_ASSIGNED]
+      isRideCancellable = maybe False (\ride -> ride.status /= SRide.CANCELLED) mbRide
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:258]: @olegromashin Even if ride status is INPROGRESS, then also we can't cancel the ride.

**Review by RitikaHanish — CHANGES_REQUESTED** [review]

File: Backend/dhall-configs/dev/dynamic-offer-driver-app.dhall (hunk 1 above)
Comment (Odeyaloleg) [reply Backend/dhall-configs/dev/dynamic-offer-driver-app.dhall]: reverted

Comment (olegromashin) [reply Backend/dhall-configs/dev/dynamic-offer-driver-app.dhall]: reverted

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs (hunk 3 above)
Comment (olegromashin) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs]: Done, please check

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnStatus.hs (hunk 2 above)
Comment (olegromashin) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/OnStatus.hs]: Done

**Review by DonDokidon — APPROVED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs [hunk 4]
```diff
@@ -326,6 +340,13 @@ onUpdate ValidatedRideCompletedReq {..} = do
           }
 onUpdate ValidatedBookingCancelledReq {..} = do
   mbRide <- QRide.findActiveByRBId booking.id
+
+  let isBookingCancellable =
+        booking.status `elem` [SRB.NEW, SRB.CONFIRMED, SRB.AWAITING_REASSIGNMENT, SRB.TRIP_ASSIGNED]
+      isRideCancellable = maybe False (\ride -> ride.status `notElem` [SRide.INPROGRESS, SRide.CANCELLED]) mbRide
+      bookingAlreadyCancelled = booking.status == SRB.CANCELLED
+  unless (isBookingCancellable || (bookingAlreadyCancelled && isRideCancellable)) $
```
Comment (roman-bodavskiy) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs:348]: These checks are already there in `validateRequest` function. So could you combine them?

Comment (olegromashin) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Beckn/OnUpdate.hs]: Done. Please, check

File: Backend/dev/migrations/rider-dashboard/0008-ride-force-sync-api-added.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
+-- for testing purposes
+INSERT INTO atlas_bpp_dashboard.access_matrix (id, role_id, api_entity, user_access_type, user_action_type)
```
[the remark this answers is not in the document]
Comment (olegromashin) [reply Backend/dev/migrations/rider-dashboard/0008-ride-force-sync-api-added.sql]: Fixed

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/Status.hs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    buildTaxiContext
+      Context.STATUS
+      messageId
+      (Just booking.transactionId)
+      bapIDs.cabs
+      bapURIs.cabs
+      (Just merchant.id.getId)
+      (Just booking.providerUrl)
+      merchant.city
+  pure $ BecknReq context $ Status.StatusMessage rideId.getId
```
Comment (roman-bodavskiy) [inline Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/Status.hs:48]: you should send bpp `rideId`, not bap `rideId`:

```
data StatusMessage = StatusMessage
  { order_id :: Text, -- bpp bookingId
    fulfillment ::  FulfillmentInfo
  }

newtype FulfillmentInfo = FulfillmentInfo
  { id :: Maybe Text -- bppRideId
  }
```

Comment (roman-bodavskiy) [inline Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/Status.hs:48]: you should send bpp `rideId`, not bap `rideId`:

```
data StatusMessage = StatusMessage
  { order_id :: Text, -- bpp bookingId
    fulfillment ::  FulfillmentInfo
  }

newtype FulfillmentInfo = FulfillmentInfo
  { id :: Text -- bppRideId (can be optional as well)
  }
```

**Review by roman-bodavskiy — COMMENTED** [review]
^

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/Status.hs (hunk 6 above)
Comment (olegromashin) [reply Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/Status.hs]: Fixed

**Review by hkmangla — APPROVED** [review]
